From de6fed308eb5e428b778001d9f5016dcd87011e1 Mon Sep 17 00:00:00 2001
From: Keerthy <j-keerthy@ti.com>
Date: Mon, 27 Mar 2017 14:00:13 +0530
Subject: [PATCH 156/205] net: wireless: ti: wlcore: sdio: fix NULL pointer
 access during suspend

Access to struct wl1271 through wl results in a NULL pointer
crash during wl1271_suspend().

wl1271_op_suspend() in drivers/net/wireless/ti/wlcore/main.c sets
wow_enabled to true always.

Fix the NULL pointer exception by assuming wake-on-wlan is always
enabled in SDIO layer too.

Signed-off-by: Keerthy <j-keerthy@ti.com>
Acked-by: Eyal Reizer <eyalr@ti.com>
Signed-off-by: Sekhar Nori <nsekhar@ti.com>
(cherry picked from commit 61f69a776c67f6ff0271354ebbd0dbf3ff894910)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/wireless/ti/wlcore/sdio.c |   38 ++++++++++++++++-----------------
 1 file changed, 18 insertions(+), 20 deletions(-)

diff --git a/drivers/net/wireless/ti/wlcore/sdio.c b/drivers/net/wireless/ti/wlcore/sdio.c
index 287023e..5bf2ed1 100644
--- a/drivers/net/wireless/ti/wlcore/sdio.c
+++ b/drivers/net/wireless/ti/wlcore/sdio.c
@@ -399,31 +399,29 @@ static int wl1271_suspend(struct device *dev)
 	/* Tell MMC/SDIO core it's OK to power down the card
 	 * (if it isn't already), but not to remove it completely */
 	struct sdio_func *func = dev_to_sdio_func(dev);
-	struct wl12xx_sdio_glue *glue = sdio_get_drvdata(func);
-	struct wl1271 *wl = platform_get_drvdata(glue->core);
 	mmc_pm_flag_t sdio_flags;
 	int ret = 0;
 
-	dev_dbg(dev, "wl1271 suspend. wow_enabled: %d\n",
-		wl->wow_enabled);
-
-	/* check whether sdio should keep power */
-	if (wl->wow_enabled) {
-		sdio_flags = sdio_get_host_pm_caps(func);
+	/**
+	 * Due to some mmc layer issues, the system automatically
+	 * powers us up on resume, which later cause issues when
+	 * we try to restore_power again explicitly.
+	 * temporarily workaround by always asking to keep power.
+	 * this is fine as driver controls the chip power anyway.
+	 */
+	sdio_flags = sdio_get_host_pm_caps(func);
 
-		if (!(sdio_flags & MMC_PM_KEEP_POWER)) {
-			dev_err(dev, "can't keep power while host "
-				     "is suspended\n");
-			ret = -EINVAL;
-			goto out;
-		}
+	if (!(sdio_flags & MMC_PM_KEEP_POWER)) {
+		dev_err(dev, "can't keep power while host is suspended\n");
+		ret = -EINVAL;
+		goto out;
+	}
 
-		/* keep power while host suspended */
-		ret = sdio_set_host_pm_flags(func, MMC_PM_KEEP_POWER);
-		if (ret) {
-			dev_err(dev, "error while trying to keep power\n");
-			goto out;
-		}
+	/* keep power while host suspended */
+	ret = sdio_set_host_pm_flags(func, MMC_PM_KEEP_POWER);
+	if (ret) {
+		dev_err(dev, "error while trying to keep power\n");
+		goto out;
 	}
 out:
 	return ret;
-- 
1.7.9.5

